### PR TITLE
Win reg stat module

### PR DESCRIPTION
**Pull request by jborean93**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
win_reg_stat

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
Added win_reg_stat module to help with querying windows registry keys and dealing with things like string expansion keys and other weird registry peculiarities. This is the new PR https://github.com/ansible/ansible-modules-extras after the repo merge

### CODE REVIEW

**Comment by jhawkesworth** [comment]
@jborean93 Hey any chance you could rebase this - I want to test it out?
Many thanks,
Jon

**Comment by jborean93** [comment]
Happy to, most of the times I have done this I seem to screw it up. Is there a dedfinitive guide on how to do this or is it just a merge with the latest commit on devel and then rebase it?

**Comment by jhawkesworth** [comment]
I am not aware of a definitive guide. 

I haven't had to do it in a while, but I think it would be a case of doing a git fetch upstream and then a git rebase upstream/devel

You might get merge conflicts which you'd have to fix manually and then commit and push.  I can't now recall if you would have to force push but github will tell you if you need to.

**Comment by jborean93** [comment]
Hey @jhawkesworth I've merged the branch with the latest devel branch in the main ansible repo. I'm still having some issues with rebasing/squashing the commits into 1 but hopefully this is what you wanted. Please let me know if you wanted anything else changed.

**Comment by jhawkesworth** [comment]
Thanks for updating this.  What I can see on github looks ok.  I don't think commits from a single contributor have to be squashed before merging, but we can tackle that at the time if it proves necessary.

Finally got round to running your tests against my test vms.

I had a failure on S2008R2 in the validation following the 

"win_reg_stat : set expected value for reg structure"

step of your integration tests.

Error was: 

fatal: [192.168.137.69]: FAILED! => {
    "assertion": "actual_short == expected",
    "changed": false,
    "evaluated_to": false,
    "failed": true
}

Its either a/ something hoky about my test VM or a difference cause by S2008R2 running WMF 3/ Powershell 3.0  (still lowest supported version in ansible).

For this reason I think this...

needs_revision

**Comment by jborean93** [comment]
Hmm I'm not sure what the issue would be, both my test VMs at home (including 2008R2) and the automated tests have worked.

I'll have a look a bit further in the coming weeks to find out what may have been the cause.

**Comment by jhawkesworth** [comment]
Thanks.

I've done a few re-runs.  S2008r2 only seems to fail first time I ran it, subsequent runs ran through ok.

Just had the failure at the same point against my win10 host - here's the context (sorry for the cow noise).  I've run 3 times against win 10 and I get the same failure in the same place.

Looking at the test and what's in the registry I think something is up with the multi part, looks like the test is expecting 
`          multi: { raw_value: ['a, b', 'c'], type: 'REG_MULTI_SZ', value: ['a, b', 'c'] }`
 but the registry has just {a, b, c} (ps output below).

```
PS C:\Users\jon> cd hkcu:
PS HKCU:\> cd Test
PS HKCU:\Test> ls


    Hive: HKEY_CURRENT_USER\Test


Name                           Property
----                           --------
nested                         string : test
                               binary : {1, 22}
                               dword  : 1
                               qword  : 1
                               multi  : {a, b, c}
                               expand : C:\WINDOWS\dir
single                         string1 :
                               string2 : abc123
                               none    : {}
                               none1   : {0}

```




failure in context below.
```
/ TASK [win_reg_stat : get known nested reg key structure for \
\ testing with quoted yaml]                                   /
 -------------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

ok: [TENSY.home] => {"changed": false, "win_reg_stat": {"exists": true, "properties": {"binary": {"raw_value": ["0x01", "0x16"], "type": "REG_BINARY", "value": [1, 22]}, "dword": {"raw_value": 1, "type": "REG_DWORD", "value": 1}, "expand": {"raw_value": "%windir%\\dir", "type": "REG_EXPAND_SZ", "value": "C:\\WINDOWS\\dir"}, "multi": {"raw_value": ["a, b", "c"], "type": "REG_MULTI_SZ", "value": ["a, b", "c"]}, "qword": {"raw_value": 1, "type": "REG_QWORD", "value": 1}, "string": {"raw_value": "test", "type": "REG_SZ", "value": "test"}}, "sub_keys": ["nest1", "nest2"]}}
ok: [WIN-12.home] => {"changed": false, "win_reg_stat": {"exists": true, "properties": {"binary": {"raw_value": ["0x01", "0x16"], "type": "REG_BINARY", "value": [1, 22]}, "dword": {"raw_value": 1, "type": "REG_DWORD", "value": 1}, "expand": {"raw_value": "%windir%\\dir", "type": "REG_EXPAND_SZ", "value": "C:\\Windows\\dir"}, "multi": {"raw_value": ["a, b", "c"], "type": "REG_MULTI_SZ", "value": ["a, b", "c"]}, "qword": {"raw_value": 1, "type": "REG_QWORD", "value": 1}, "string": {"raw_value": "test", "type": "REG_SZ", "value": "test"}}, "sub_keys": ["nest1", "nest2"]}}
ok: [WIN-2K82] => {"changed": false, "win_reg_stat": {"exists": true, "properties": {"binary": {"raw_value": ["0x01", "0x16"], "type": "REG_BINARY", "value": [1, 22]}, "dword": {"raw_value": 1, "type": "REG_DWORD", "value": 1}, "expand": {"raw_value": "%windir%\\dir", "type": "REG_EXPAND_SZ", "value": "C:\\Windows\\dir"}, "multi": {"raw_value": ["a, b", "c"], "type": "REG_MULTI_SZ", "value": ["a, b", "c"]}, "qword": {"raw_value": 1, "type": "REG_QWORD", "value": 1}, "string": {"raw_value": "test", "type": "REG_SZ", "value": "test"}}, "sub_keys": ["nest1", "nest2"]}}
 ____________________________________________________________
< TASK [win_reg_stat : set expected value for reg structure] >
 ------------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

ok: [WIN-2K82] => {"ansible_facts": {"expected": {"changed": false, "win_reg_stat": {"exists": true, "properties": {"binary": {"raw_value": ["0x01", "0x16"], "type": "REG_BINARY", "value": [1, 22]}, "dword": {"raw_value": 1, "type": "REG_DWORD", "value": 1}, "expand": {"raw_value": "%windir%\\dir", "type": "REG_EXPAND_SZ", "value": "C:\\Windows\\dir"}, "multi": {"raw_value": ["a, b", "c"], "type": "REG_MULTI_SZ", "value": ["a, b", "c"]}, "qword": {"raw_value": 1, "type": "REG_QWORD", "value": 1}, "string": {"raw_value": "test", "type": "REG_SZ", "value": "test"}}, "sub_keys": ["nest1", "nest2"]}}}, "changed": false}
ok: [TENSY.home] => {"ansible_facts": {"expected": {"changed": false, "win_reg_stat": {"exists": true, "properties": {"binary": {"raw_value": ["0x01", "0x16"], "type": "REG_BINARY", "value": [1, 22]}, "dword": {"raw_value": 1, "type": "REG_DWORD", "value": 1}, "expand": {"raw_value": "%windir%\\dir", "type": "REG_EXPAND_SZ", "value": "C:\\Windows\\dir"}, "multi": {"raw_value": ["a, b", "c"], "type": "REG_MULTI_SZ", "value": ["a, b", "c"]}, "qword": {"raw_value": 1, "type": "REG_QWORD", "value": 1}, "string": {"raw_value": "test", "type": "REG_SZ", "value": "test"}}, "sub_keys": ["nest1", "nest2"]}}}, "changed": false}
ok: [WIN-12.home] => {"ansible_facts": {"expected": {"changed": false, "win_reg_stat": {"exists": true, "properties": {"binary": {"raw_value": ["0x01", "0x16"], "type": "REG_BINARY", "value": [1, 22]}, "dword": {"raw_value": 1, "type": "REG_DWORD", "value": 1}, "expand": {"raw_value": "%windir%\\dir", "type": "REG_EXPAND_SZ", "value": "C:\\Windows\\dir"}, "multi": {"raw_value": ["a, b", "c"], "type": "REG_MULTI_SZ", "value": ["a, b", "c"]}, "qword": {"raw_value": 1, "type": "REG_QWORD", "value": 1}, "string": {"raw_value": "test", "type": "REG_SZ", "value": "test"}}, "sub_keys": ["nest1", "nest2"]}}}, "changed": false}
 _____________________________________
< TASK [win_reg_stat : validate test] >
 -------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

ok: [WIN-2K82] => {
    "changed": false,
    "msg": "All assertions passed"
}
fatal: [TENSY.home]: FAILED! => {
    "assertion": "actual_short == expected",
    "changed": false,
    "evaluated_to": false,
    "failed": true
}
ok: [WIN-12.home] => {
    "changed": false,
    "msg": "All assertions passed"
}

```

For info, in regedit I can see the multi value as 'a, b c'

I am running the ansible controller under bash for ubuntu for windows on the same win10 machine that I am using for a target.  The 2008r2/2012r2 machines are running in virtualbox, if that's any help.

Jon

**Comment by jborean93** [comment]
Thanks for the info @jhawkesworth I'll have a play around locally and see if I can replicate the same issue and come up with a fix. Weird that this is happening but I am sure if it some sort of Windows quirk.

**Comment by jborean93** [comment]
@jhawkesworth I have run this through more than a dozen times and I cannot get the same error you are coming across. Are you able to make sure that the registry key `HKEY_CURRENT_USER:\TEST` has been deleted from your test box before trying again?

**Comment by jhawkesworth** [comment]
@jborean93 - i tried again earlier.  I cleaned the HKCU:\TEST key and ran against same set of machines S2008r2, s2012r2 (both running in virtualbox) and win10 (host pc) and again it failed on the windows 10 host at the same point.
It is actually when it is checking the asserts that it fails, rather than during any powershell operations on the windows machines.  I am speculating that it is perhaps a side effect of me running the playbook from the Ubuntu installation hosted under Windows Subsystem for Linux (WSL) running on the same win10 host that is being targetted.  I will try running from a linux vm and report back.

**Comment by jborean93** [comment]
I believe I also checked the output message you showed and the 2 json structures were the same for actual and expected so I was confused (I could have been tired though). Just to confirm I am using WSL as well but not on the same host so I will try that later.

**Comment by jhawkesworth** [comment]
just tried running the integration test from lubuntu vm but got the same results.
Will try on a different machine.  I really want to get this merged but also really want to understand why the assert is tripping up in this way too.

**Comment by jborean93** [comment]
I think I figured it out, it isn't the multi string that is failing, this is your actual return result

`{"changed": false, "win_reg_stat": {"exists": true, "properties": {"binary": {"raw_value": ["0x01", "0x16"], "type": "REG_BINARY", "value": [1, 22]}, "dword": {"raw_value": 1, "type": "REG_DWORD", "value": 1}, "expand": {"raw_value": "%windir%\\dir", "type": "REG_EXPAND_SZ", "value": "C:\\WINDOWS\\dir"}, "multi": {"raw_value": ["a, b", "c"], "type": "REG_MULTI_SZ", "value": ["a, b", "c"]}, "qword": {"raw_value": 1, "type": "REG_QWORD", "value": 1}, "string": {"raw_value": "test", "type": "REG_SZ", "value": "test"}}, "sub_keys": ["nest1", "nest2"]}}`

This is the expected return result
`{"changed": false, "win_reg_stat": {"exists": true, "properties": {"binary": {"raw_value": ["0x01", "0x16"], "type": "REG_BINARY", "value": [1, 22]}, "dword": {"raw_value": 1, "type": "REG_DWORD", "value": 1}, "expand": {"raw_value": "%windir%\\dir", "type": "REG_EXPAND_SZ", "value": "C:\\Windows\\dir"}, "multi": {"raw_value": ["a, b", "c"], "type": "REG_MULTI_SZ", "value": ["a, b", "c"]}, "qword": {"raw_value": 1, "type": "REG_QWORD", "value": 1}, "string": {"raw_value": "test", "type": "REG_SZ", "value": "test"}}, "sub_keys": ["nest1", "nest2"]}}`

The only difference I can see is the REG_EXPAND_SZ entry where the value changes from `C:\\Windows\\dir` to `C:\\WINDOWS\\dir`. Are you able to check if the `%windir%` system variable is set differently on your Win10 host compared to the others. I don't have access to a Win 10 host right now to see if they are set differently.

**Comment by jhawkesworth** [comment]
Yep you are right, its C:\WINDOWS on my windows 10 host and C:\Windows on the server2012r2 and s2008r2.
I checked a couple of machines at work and its the same there.  Mystery solved!
If you can figure a way to fix up the test I can't see a barrier to getting this merged in.

**Comment by jborean93** [comment]
@jhawkesworth the tests have been updated and I have tested it locally with both uppercase and normal style, should be good for you to test again.

**Comment by jhawkesworth** [comment]
Great, thank you.

I just re-pulled and re-rand the tests are running through for me now.  Run them twice without error against same 8 / 10 / 12 test machines, so I am happy for this to go in.

Thank you so much for your persistence with this module!

shipit

**Comment by jborean93** [comment]
ready_for_review

**Comment by jborean93** [comment]
@jhawkesworth thanks for your review let me know if you have anything you wish for me to review, I know I've still got some more to go. Are you able to put in shipit as the docker tests failed due to a host not coming online, it's since run correctly but lost the community review.

**Comment by jhawkesworth** [comment]
to keep the bot happy

shipit

**Comment by jhawkesworth** [comment]
shippable tests seem to be failing in tests for 'yum' - unrelated to this PR

**Comment by jborean93** [comment]
Still having issues, sorry for the spam
ready_for_review

**Comment by mattclay** [comment]
bot_status

**Comment by dagwieers** [comment]
Please rebase (and please add them alphabetically), then shipit !

**Comment by jborean93** [comment]
@dagwieers I have rebased from the latest devel branch, hopefully Shippable doesn't fail while waiting for the test host.

**Comment by dagwieers** [comment]
Don't mind the Shippable output... It's  been unreliable the past few days, to the point that I assume I am right and Shippable is wrong :-D

**Comment by dagwieers** [comment]
shipit

**Comment by jborean93** [comment]
Actually I did my rebase wrong, suprising it is a problem for me and not shippable :). Will fix it up in a few minutes.

**Comment by jborean93** [comment]
@dagwieers I've fixed up my mistake and shippable is now passing.

**Comment by dagwieers** [comment]
@jborean93 And you've been lucky that the tests (https://github.com/ansible/ansible/commit/189fee48539b405357cf43145c4c4f96923ac006) were fixed as well ;-)

I'd say shipit !

**Review by dagwieers — APPROVED** [review]

**Comment by dagwieers** [comment]
Noticed win_reg_stat was not added alphabetically to CHANGELOG.md, but I am sure someone will fix that eventually !

**Comment by jborean93** [comment]
oops didn't know it was in alphabetical order.

**Comment by dagwieers** [comment]
You are not alone, openstack and aws have issues now too.
If you are listing things (even in code) and the order has no meaning, alphabetical order is always preferred.
Because one can easily compare lists, but also because it is easier to find something you are looking for.

File: lib/ansible/modules/windows/win_reg_stat.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_reg_stat.ps1]: TODO item is no longer needed.

Comment (jborean93) [reply lib/ansible/modules/windows/win_reg_stat.ps1]: You are right, will change and fix up the changelog as well.

**Comment by jborean93** [comment]
@dagwieers those changes have been made, thanks for your review

**Comment by jhawkesworth** [comment]
bot_status

**Comment by dagwieers** [comment]
shipit

**Comment by jhawkesworth** [comment]
since there have been revisions, looks like I need to say the magic word again

shipit

**Comment by dagwieers** [comment]
Is that true ? Maybe I need a bot for that...

**Comment by jborean93** [comment]
@gundalow I'm not sure if you're the right person but I have seen you being mentioned before, is this meant to auto merge now that there are 2 shipits?

**Comment by mattclay** [comment]
bot_status

**Comment by dagwieers** [comment]
shipit

**Comment by jhawkesworth** [comment]
bot_status

**Comment by dagwieers** [comment]
shipit

**Comment by jborean93** [comment]
bot_broken

**Comment by jhawkesworth** [comment]
Obviously we all want the bot fixed but appealing to @nitzmahone for a final review on this one.

@jctanner - not sure if you are maintaining the bot these days.  Just wondering if it is possible that the  fact I am in external commiters group might be confusing the count of ship_it?

**Comment by dagwieers** [comment]
@jhawkesworth Your shipit did count, it was mine  that was not counted.

**Review by nitzmahone — APPROVED** [review]

**Comment by nitzmahone** [comment]
Thanks!

**Comment by jhawkesworth** [comment]
Great, glad this is in, gonna be so useful.

**Comment by jborean93** [comment]
@gundalow is there anything that needs to be done to assign maintainers like you did with the previous module?

**Comment by jhawkesworth** [comment]
@jborean93 - I believe you can make a PR on https://github.com/ansible/ansibullbot/ (its just modifying maintainers.txt).  Once you've created the PR ask on #ansible-devel irc channel and someone will merge it.

**Comment by jborean93** [comment]
Thanks @jhawkesworth I'll raise a PR there, wasn't sure what the correct protocol was.

**Pull request merged**
